Gems/Compression/Code/Source/Tools/CompressionFactoryImpl.h

```diff
@@ -21,7 +21,8 @@ namespace Compression
 
         ~CompressionFactoryImpl();
 
-        AZStd::span<ICompressionInterface* const> GetCompressionInterfaces() const override;
+        AZStd::span<AZStd::unique_ptr<ICompressionInterface> const> GetCompressionInterfaces() const override;
+        void VisitCompressionInterfaces(const VisitCompressionInterfaceCallback&) const override;
         bool RegisterCompressionInterface(AZStd::unique_ptr<ICompressionInterface>&&) override;
         bool UnregisterCompressionInterface(CompressionAlgorithmId) override;
         ICompressionInterface* FindCompressionInterface(CompressionAlgorithmId) const override;
```

Comment:
you return a span of unique_ptr but return a raw pointer in this case. would it be more consistent to use a visitor interface:? I see you already have that on line 25?

Reply:
You are right, I will remove the GetCompressionInterfaces/GetDecompressionInterfaces functions since it is a strange to return span of vector of unique_ptrs.

Reply:
Updated.
Removing the `GetCompressionInterfaces` API also has the benefit of not exposing the implementation detail that the compression interfaces are stored in unique_ptrs.

Reply:
would it make sense to do the same treatment for FindCompressionInterface

```
void VisitCompressionInterfaces(CompressionAlgorithmId, const VisitCompressionInterfaceCallback&) const override;
```

Reply:
Not really for FindCompressionInterface, since this will be a high frequency function to be called to get a pointer to a specific CompressionInterface.

The `VisitCompressionInterfaces` has less uses and will probably be called less frequently.

Furthermore `FindCompressionInterface` doesn't need to allocate to return a pointer and it doesn't enforce a particular data structure for the CompressionFactoryImpl, unlike the former `GetCompressionInterfaces` function, which required that internal data structure of the CompressionFactoryImpl had to store a vector of raw CompressionInterface pointers.

That is the reason for the Visit function there.
It allows users to access the CompressionInterfaces without requiring any dynamic memory allocations.
Furthermore it allows the user to choose the container to populate with the CompressionInterfaces if they so choose.